OTAnalytics/plugin_ui/customtkinter_gui/frame_canvas.py

```diff
@@ -118,42 +135,89 @@ def __init__(self, canvas: CanvasBackground):
+    ) -> None:
         for observer in self._observers:
-            observer.update(coordinates, event_type)
+            observer.update(coordinates, event_type, key)
+
+    def _on_left_button_down(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, LEFT_BUTTON_DOWN)
+
+    def _on_left_button_up(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+
+        self._notify_observers(coordinates, LEFT_BUTTON_UP)
+
+    def _on_right_button_up(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, RIGHT_BUTTON_UP)
+
+    def _on_mouse_motion(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, MOTION)
+
+    def _on_motion_while_left_button_down(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, MOTION_WHILE_LEFT_BUTTON_DOWN)
+
+    def _on_mouse_leaves_canvas(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, LEAVE_CANVAS)
+
+    def _on_mouse_enters_canvas(self, event: Any) -> None:
+        self._canvas.focus_set()
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, ENTER_CANVAS)
+
+    def _on_left(self, event: Any) -> None:
+        coordinates = self._get_mouse_coordinates(event)
+        self._notify_observers(coordinates, LEFT_KEY)
 
-    def on_left_mousebutton_up(self, event: Any) -> None:
+    def _on_right(self, event: Any) -> None:
         coordinates = self._get_mouse_coordinates(event)
-        self._notify_observers(coordinates, "left_mousebutton_up")
+        self._notify_observers(coordinates, RIGHT_KEY)
 
-    def on_right_mousebutton_up(self, event: Any) -> None:
+    def _on_return(self, event: Any) -> None:
         coordinates = self._get_mouse_coordinates(event)
-        self._notify_observers(coordinates, "right_mousebutton_up")
+        self._notify_observers(coordinates, RETURN_KEY)
 
-    def on_mouse_motion(self, event: Any) -> None:
+    def _on_delete(self, event: Any) -> None:
         coordinates = self._get_mouse_coordinates(event)
-        self._notify_observers(coordinates, "mouse_motion")
+        self._notify_observers(coordinates, DELETE_KEYS)
 
-    def on_return(self, event: Any) -> None:
+    def _on_escape(self, event: Any) -> None:
         coordinates = self._get_mouse_coordinates(event)
-        self._notify_observers(coordinates, "return")
+        self._notify_observers(coordinates, ESCAPE_KEY)
 
-    def on_escape(self, event: Any) -> None:
+    def _on_other_key_up(self, event: Any) -> None:
+        if event.keysym in KEY_SYMBOLS:
+            return
+        key = event.char
```

Comment:
Because I thought it was best to bind functional keys like `return`, `escape` or `del` "explicitly" and bind "normal" keys like `a`, `+` or `!` "generically". Or should we better treat all keys the same?